File: message/message.go
package message

import (
	"bytes"
	"fmt"
	"strconv"
	"strings"
)

// Message ...
type Message struct {
	ID        string
	Role      string
	Seq       int
	IsEnd     bool
	Status    int
	MediaType string
	Encoding  string
	Data      []byte
}

// ParseMessage expects arg byteSlice to be of format: <header>;<data>
func ParseMessage(byteSlice []byte) (*Message, error) {
	semicolonIndex := bytes.IndexRune(byteSlice, ';')
	if semicolonIndex == -1 {
		err := fmt.Errorf("Cannot parse message: %v", string(byteSlice))
		return nil, err
	}
	msgString := string(byteSlice[:semicolonIndex])
	comps := strings.Split(msgString, ",")
	return &Message{
		ID:        comps[0],
		Role:      comps[1],
		Seq:       parseInt(comps[2]),
		IsEnd:     parseBool(comps[3]),
		Status:    parseInt(comps[4]),
		MediaType: comps[5],
		Encoding:  comps[6],
		Data:      byteSlice[semicolonIndex+1:],
	}, nil
}

func (m *Message) headerBytes() []byte {
	comps := []string{
		m.ID,
		m.Role,
		formatInt(m.Seq),
		strconv.FormatBool(m.IsEnd),
		formatInt(m.Status),
		m.MediaType,
		m.Encoding,
	}
	str := strings.Join(comps, ",")
	bts := []byte(str)
	bts = append(bts, ';')
	return bts
}

// Envelope returns length-prefixed envelope of format: <byte-length-header-data>:<header>;<data>
func (m *Message) Envelope() *Envelope {
	return NewEnvelope(m.headerBytes(), m.Data)
}

func parseInt(comp string) int {
	value, err := strconv.ParseInt(comp, 10, 32)
	catch(err)
	return int(value)
}

func formatInt(value int) string {
	return strconv.FormatInt(int64(value), 10)
}

func parseBool(comp string) bool {
	value, err := strconv.ParseBool(comp)
	catch(err)
	return value
}

func catch(err error) {
	if err != nil {
		panic(err)
	}
}

File: node-approx/message.js
class Message {
  constructor({id, role, seq, isEnd, status, mediaType, encoding, data}) {
    this.id = id || '';
    this.role = role || '';
    this.seq = typeof seq === 'number' ? seq : 0;
    this.isEnd = typeof isEnd === 'boolean' ? isEnd : true;
    this.status = typeof status === 'number' ? status : 0;
    this.mediaType = mediaType || '';
    this.encoding = encoding || '';
    if (data) {
      this.data = data instanceof Buffer ? data : Buffer.from(data);
    } else {
      this.data = Buffer.from('');
    }
  }
  static from({id, role, seq, isEnd, status, mediaType, encoding, data}) {
    return new Message({id, role, seq, isEnd, status, mediaType, encoding, data});
  }
  static parse({buffer}) {
    const semicolonIndex = buffer.indexOf(';');
    if (semicolonIndex === -1) throw new Error(`Cannot parse message: ${buffer.toString('utf8')}`);
    const headerBuff = buffer.slice(0, semicolonIndex);
    const dataBuff = buffer.slice(semicolonIndex + 1);
    const comps = headerBuff.toString('utf8').split(',');
    const [id, role, seqStr, isEndStr, statusStr, mediaType, encoding] = comps;
    const data = encoding === 'base64'
      ? Buffer.from(dataBuff.toString('utf8'), 'base64')
      : dataBuff;
    const seq = Number.parseInt(seqStr);
    const isEnd = Message._parseBool(isEndStr);
    const status = Number.parseInt(statusStr);
    return new Message({id, role, seq, isEnd, status, mediaType, encoding, data}); 
  }
  static _parseBool(strBool, fallback) {
    if (typeof strBool === 'boolean') return strBool;
    strBool = strBool.toUpperCase();
    let value;
    switch (strBool) {
      case 'TRUE':
      case 'T':
      case '1':
        value = true;
        break;
      case 'FALSE':
      case 'F':
      case '0':
        value = false;
        break;
      default:
        value = fallback;
        break;
    }
    return value;
  }
  _headerBuffer() {
    const comps = [
      this.id,
      this.role,
      this.seq,
      this.isEnd,
      this.status,
      this.mediaType,
      this.encoding,
    ];
    const str = comps.join(',');
    return Buffer.from(`${str};`, 'utf8');
  }
  envelope() {
    const headerBuff = this._headerBuffer();
    const length = headerBuff.length + this.data.length;
    const lengthBuff = Buffer.from(`${length}:`);
    return Buffer.concat(
      [lengthBuff, headerBuff, this.data], 
      lengthBuff.length + length,
    );
  }
}

module.exports = Message;
